crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+                    }
+                } else {
+                    lines.push(DunderAllLine::OneOrMoreItems(LineWithItems::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range,
+                    )));
+                }
+                break;
+            }
+            Tok::NonLogicalNewline => {
+                if items_in_line.is_empty() {
+                    if let Some(comment) = comment_in_line {
+                        lines.push(DunderAllLine::JustAComment(LineWithJustAComment::new(
+                            comment, range,
+                        )));
+                        comment_in_line = None;
+                    }
+                } else {
+                    lines.push(DunderAllLine::OneOrMoreItems(LineWithItems::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range,
+                    )));
+                    comment_in_line = None;
+                    items_in_line.clear();
+                }
+            }
+            Tok::Comment(_) => comment_in_line = Some(subrange),
+            Tok::String { value, .. } => items_in_line.push((value, subrange)),
+            Tok::Comma => continue,
+            _ => return None,
+        }
+    }
+    Some(lines)
+}
+
+#[derive(Debug)]
+struct LineWithJustAComment(TextRange);
+
+impl LineWithJustAComment {
+    fn new(comment_range: TextRange, total_dunder_all_range: TextRange) -> Self {
+        Self(comment_range + total_dunder_all_range.start())
+    }
+}
+
+#[derive(Debug)]
+struct LineWithItems {
+    items: Vec<(String, TextRange)>,
+    comment_range: Option<TextRange>,
+}
+
+impl LineWithItems {
+    fn new(
+        items: &[(String, TextRange)],
+        comment_range: Option<TextRange>,
+        total_dunder_all_range: TextRange,
+    ) -> Self {
+        assert!(
+            !items.is_empty(),
+            "Use the 'JustAComment' variant to represent lines with 0 items"
+        );
+        let offset = total_dunder_all_range.start();
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
+                .collect(),
+            comment_range: comment_range.map(|c| c + offset),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DunderAllLine {
+    JustAComment(LineWithJustAComment),
+    OneOrMoreItems(LineWithItems),
+}
+
+fn collect_dunder_all_items(lines: &[DunderAllLine]) -> Vec<DunderAllItem> {
+    // Given data on each line in `__all__`, group lines together into "items".
+    // Each item contains exactly one element,
+    // but might contain multiple comments attached to that element
+    // that must move with the element when `__all__` is sorted.
+    let mut all_items = vec![];
+    let mut this_range = None;
+    for line in lines {
+        match line {
+            DunderAllLine::JustAComment(LineWithJustAComment(comment_range)) => {
```

Comment:
What happens with `JustAComment` if it is the last line? 

```python
__all__ = [
    'test',
    # comment
]
```

I'm asking because `this_range` then never gets added to all items. Will it be part of the prelude or is there an invariant that prevents that `LineWithJustAComment` can be the last line?

Reply:
Any comments after the last item are discarded here, but are added back as part of the postlude in `into_sorted_source_code()` (which used to be called `construct_sorted_all()`; I just renamed it)

Reply:
Added some more docs about this; is this clearer now?

Reply:
Thanks. That helps